### PR TITLE
skip nullable iterables in `avoid_function_literals_in_foreach_calls`

### PR DESCRIPTION
**Pull request by pq**
Fixes #https://github.com/dart-lang/linter/issues/2751

/cc @bwilkerson

### CODE REVIEW

File: test_data/rules/avoid_function_literals_in_foreach_calls.dart [hunk 1]
```diff
@@ -9,12 +9,14 @@ class Person {
 }
 
 void main() {
-  Iterable<String> people;
+  Iterable<String> people = [];
```
Comment (bwilkerson) [inline test_data/rules/avoid_function_literals_in_foreach_calls.dart:12]: Should we add something to the test framework to fail a test if there are any diagnostics other than the lints being tested for? Seems like this is something that should have been caught when switching to analyze test code in null safety by default.

File: test_data/rules/avoid_function_literals_in_foreach_calls.dart [hunk 2]
```diff
@@ -9,12 +9,14 @@ class Person {
...
 
-  for (var person in people) { // OK
+  for (var person in people) {
+    // OK
     print('$person!');
   }
-  people.forEach((person) { // LINT
+  people.forEach((person) // LINT
```
Comment (bwilkerson) [inline test_data/rules/avoid_function_literals_in_foreach_calls.dart:18]: I'm not sure of the value of moving the opening bracket, but ok.

File: lib/src/rules/avoid_function_literals_in_foreach_calls.dart [hunk 3]
```diff
@@ -31,6 +33,27 @@ people.forEach(print);
...
+    } else if (exp is PropertyAccess) {
+      exp = exp.target;
+    }
+  }
+  return false;
+}
+
+bool _isNonNullableIterable(DartType? type) =>
+    type != null &&
+    type.nullabilitySuffix != NullabilitySuffix.question &&
```
Comment (bwilkerson) [inline lib/src/rules/avoid_function_literals_in_foreach_calls.dart:54]: I'm not sure how this works for `dynamic`. Consider
```
void f(dynamic iter) => iter?.forEach(...);
```
Is the suffix always a question even when it isn't explicitly given?

I'll just note that an equivalent test might be to look for the `?.` operator on the method invocation.

**Review by bwilkerson — APPROVED** [review]
LGTM assuming it works with `dynamic`.

File: test_data/rules/avoid_function_literals_in_foreach_calls.dart (hunk 2 above)
Comment (pq) [reply test_data/rules/avoid_function_literals_in_foreach_calls.dart:18]: It's a formatting thing.  The formatter moves the comment if it's after the brace, which breaks the expectation.

File: lib/src/rules/avoid_function_literals_in_foreach_calls.dart (hunk 3 above)
Comment (pq) [reply lib/src/rules/avoid_function_literals_in_foreach_calls.dart:54]: > Is the suffix always a question even when it isn't explicitly given?

I **_believe_** so?  (@scheglov may knot for sure.)

In any event, I added a `dynamic` test case and it seems to work as written.

Thanks for the catch!

File: test_data/rules/avoid_function_literals_in_foreach_calls.dart (hunk 1 above)
Comment (pq) [reply test_data/rules/avoid_function_literals_in_foreach_calls.dart:14]: Good question. We're currently ignoring diagnostics and I suspect there are a ton.

https://github.com/dart-lang/linter/issues/2760

File: lib/src/rules/avoid_function_literals_in_foreach_calls.dart (hunk 3 above)
Comment (scheglov) [reply lib/src/rules/avoid_function_literals_in_foreach_calls.dart:54]: The type `dynamic` is always nullable, and it does not have any `nullabilitySuffix`, it is always `NullabilitySuffix.none`. Note that the suffix is not the same as nullability, e.g. `FutureOr<int?>` is nullable.

**Pull request merged**